File: 0041-first-missing-positive/solution.java
class Solution {
    public int firstMissingPositive(int[] nums) {
        cyclicsort(nums);
        for(int i=0;i<nums.length;i++){
            if(nums[i]==i+1){
                continue;
            }
            else{
                return i+1;
            }
        }
        return nums.length+1;
    }
    void cyclicsort(int[] nums){
        int i=0;
        while(i<nums.length){
            int correct=nums[i]-1;
            if(nums[i]>0 && nums[i]<nums.length && nums[i]!=nums[correct]){
                int temp=nums[correct];
                nums[correct]=nums[i];
                nums[i]=temp;
            }
            else{
                i++;
            }
        }
    }

}

File: 0041-first-missing-positive/solution.py
class Solution:
    def firstMissingPositive(self, arr):
        n=len(arr)
        for i in range(n):
            if(arr[i]<=0 or arr[i]>n):
                arr[i]=n+1
        for i in range(n):
            val=abs(arr[i])
            if(val>=1 and val<=n):
                if(arr[val-1]>0):
                    arr[val-1]=-arr[val-1]
        for i in range(n):
            if(arr[i]>0):
                return i+1
        return n+1
        
